clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,299 @@
+		".zip":    {},
+		".gz":     {},
+		".tar.gz": {},
+		".z":      {},
+		".bz2":    {},
+	}
+}
+
+type decompressor struct {
+	metrics   *Metrics
+	logger    log.Logger
+	handler   api.EntryHandler
+	positions positions.Positions
+
+	path string
+
+	posAndSizeMtx sync.Mutex
+	stopOnce      sync.Once
+
+	running *atomic.Bool
+	posquit chan struct{}
+	posdone chan struct{}
+	done    chan struct{}
+
+	decoder *encoding.Decoder
+
+	position int64
+	size     int64
+}
+
+func newDecompressor(metrics *Metrics, logger log.Logger, handler api.EntryHandler, positions positions.Positions, path string, encodingFormat string) (*decompressor, error) {
+	logger = log.With(logger, "component", "decompressor")
+
+	pos, err := positions.Get(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "get positions")
+	}
+
+	var decoder *encoding.Decoder
+	if encodingFormat != "" {
+		level.Info(logger).Log("msg", "decompressor will decode messages", "from", encodingFormat, "to", "UTF8")
+		encoder, err := ianaindex.IANA.Encoding(encodingFormat)
+		if err != nil {
+			return nil, errors.Wrap(err, "error doing IANA encoding")
+		}
+		decoder = encoder.NewDecoder()
+	}
+
+	decompressor := &decompressor{
+		metrics:   metrics,
+		logger:    logger,
+		handler:   api.AddLabelsMiddleware(model.LabelSet{FilenameLabel: model.LabelValue(path)}).Wrap(handler),
+		positions: positions,
+		path:      path,
+		running:   atomic.NewBool(false),
+		posquit:   make(chan struct{}),
+		posdone:   make(chan struct{}),
+		done:      make(chan struct{}),
+		position:  pos,
+		decoder:   decoder,
+	}
+
+	go decompressor.readLines()
+	go decompressor.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompressor, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompressor.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
+	ext := filepath.Ext(f.Name())
+	var decompressLib string
+
+	if strings.Contains(ext, "gz") { // .gz, .tar.gz
+		decompressLib = "compress/gzip"
+		reader, err = gzip.NewReader(f)
+	} else if ext == ".z" {
+		decompressLib = "compress/zlib"
+		reader, err = zlib.NewReader(f)
+	} else if ext == ".zip" {
+		decompressLib = "compress/flate"
+		reader = flate.NewReader(f)
+	} else if ext == ".bz2" {
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Debug(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	supportedExtsList := strings.Builder{}
+	for ext := range supportedCompressedFormats() {
+		supportedExtsList.WriteString(ext)
+	}
+	return nil, fmt.Errorf("file %q has unsupported extension, it has to be one of %q", f.Name(), supportedExtsList.String())
+}
+
+func (t *decompressor) updatePosition() {
```

Comment:
What is the motivation behind this? It's been a while but updating the position didn't have to be async.

Reply:
hmmm not sure, but I can confirm I had issues without this when using multiple clients. The other targets are also updating the positions async.